File: py/pe21.py
#! /usr/bin/env python3

from sys import argv
from pe import spd
 
def pe21(limit):
    """
    Evaluate the sum of all amicable pairs under 10000.
    """
    ami = []
    s = 0
    for i in range(2, limit + 1):
        t = spd(i)
        if i < t and i == spd(t):
            s += i + t
            ami.append((i, t))
    print(s, ami)

def main():
    argc = len(argv)
    if argc > 1:
        try:
            n = int(argv[1])
        except ValueError:
            print("  python pe21.py [integer]")
            return
    else:
        n = 10000

    pe21(n)

if __name__ == "__main__":
    main()

File: cpp/src/pe21.cc
#define _CRT_SECURE_NO_WARNINGS 1
#include <pe21.h>
#include <spd.h>

#include <iostream>

using namespace std;

static bool debug_on = false;

void pe21(uint32_t n) {
  uint32_t sum = 0;
  
  for (uint32_t i = 2; i < n + 1; ++i) {
    uint32_t t = spd(i);
    if (i < t && i == spd(t)) {
      if (debug_on) {
        cout << "(" << i << ", " << t << ")" << endl;
      }
      sum += i + t;
    }
  }
  
  cout << sum << endl;
}

int pe21_main(void) {
  for (;;) {
    uint32_t n;
    cout << "> ";
    cin >> n;
    if (n < 1) {
      break;
    }
    pe21(n);
  }

  return 0;
}
